thread_timeout/__init__.py

```diff
@@ -101,84 +103,145 @@ class NotKillExecTimeoutException(ExecTimeoutException):
 
 
 def thread_timeout(delay, kill=True, kill_wait=0.1):
-    def generate_decorator(function):
-
-        def decorator(*args, **kwargs):
-            result = [None]
-
-            def inner_worker():
-                result[0] = function(*args, **kwargs)
-            thread = threading.Thread(target=inner_worker)
-            thread.daemon = True
-            thread.start()
-            thread.join(delay)
-            if thread.isAlive():
-                    if not kill:
-                        raise NotKillExecTimeoutException(
-                            "Timeout and no kill attempt")
-                    _kill_thread(thread)
-                    time.sleep(kill_wait)
-                    if thread.isAlive():
-                        raise FailedKillExecTimeoutException(
-                            "Timeout, thread refuses to die in %s seconds" %
-                            kill_wait)
-                    else:
-                        raise KilledExecTimeoutException(
-                            "Timeout and thread was killed")
-            return result[0]
-        return decorator
-    return generate_decorator
-
-if __name__ == "__main__":
-    print ("Running tests")
-
+    @wrapt.decorator
+    def wrapper(wrapped, instance, args, kwargs):
+        queue = Queue()
+
+        def inner_worker():
+            result = wrapped(*args, **kwargs)
+            queue.put(result)
+        thread = threading.Thread(target=inner_worker)
+        thread.daemon = True
+        thread.start()
+        thread.join(delay)
+        if thread.isAlive():
+            if not kill:
+                raise NotKillExecTimeoutException(
+                    "Timeout and no kill attempt")
+            _kill_thread(thread)
+            time.sleep(kill_wait)
+            ## FIXME isAlive is giving fals positive results
```

Comment:
Please remove this from this pull and make it separate issue. It worked on my test machine and it needs separate discussion.